Amazon.Ion.ObjectMapper/IonSerializer.cs

```diff
@@ -103,15 +103,37 @@ public object Create(IonSerializationOptions options, IIonReader reader, Type ta
             if (annotations.Length > 0)
             {
                 var typeName = annotations[0];
-                var assemblyName = options.AnnotatedTypeAssemblies.First(a => Type.GetType(FullName(typeName, a)) != null);
-                return Activator.CreateInstance(Type.GetType(FullName(typeName, assemblyName)));
+                Type typeToCreate = null;
+
+                if (options.AnnotatedTypeAssemblies != null)
+                {
+                    foreach (string assemblyName in options.AnnotatedTypeAssemblies)
+                    {
+                        if ((typeToCreate = Type.GetType(FullName(typeName, assemblyName))) != null)
+                        {
+                            break;
+                        }
+                    }
+                }
+                else
+                {
+                    typeToCreate = AppDomain.CurrentDomain
+                        .GetAssemblies()
+                        .SelectMany(x => x.GetTypes())

```

Comment:
This is going to be really slow, and potentially use a lot of memory. I think, per assembly, we need to try to find the type, not copy all the types from all the assemblies into an array and then search the array.

Reply:
Revised